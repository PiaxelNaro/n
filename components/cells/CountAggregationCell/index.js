// @flow
import React from 'react';
import styled from 'styled-components';
import { translate, sprintf, type I18nType } from '../../../modules/i18n';
import Overlay from '../../Overlay';

const Link = styled.a`
    color: #102235;
`;

const ListItem = styled.div`
    margin: 10px;
    font-size: 13px;
    max-width: 180px;
`;

const _CountAggregationOverlay = ({
    count,
    link,
    items,
    i18n,
    Row,
}: {
    count: number,
    link: string,
    items: Array<Object>,
    i18n: I18nType,
    Row: Class<React$Component<void, any, any>>,
}) => {
    let moreLink;
    if (count > 10) {
        moreLink = (
            <ListItem>
                <Link href={link}>
                    {sprintf(i18n.gettext(`+ %d more`), count)}
                </Link>
            </ListItem>
        );
    }

    // NOTE: We're using index as keys because the API doesn't give us ids. We should fix this
    /* eslint-disable react/no-array-index-key */
    return (
        <Overlay showArrow>
            {items.map((item: { name: string }, i) =>
                <ListItem key={i}>
                    <Row {...item} />
                </ListItem>
            )}
            {moreLink}
        </Overlay>
    );
    /* eslint-enable react/no-array-index-key */
};

const CountAggregationOverlay = translate(_CountAggregationOverlay);

const CountAggregationCell = ({
    count,
    link,
    items,
    Row,
}: {
    count: number,
    link: string,
    items: Array<Object>,
    Row: Class<React$Component<void, any, any>>,
}) =>
    <Overlay.Trigger
        trigger="hover"
        getOverlay={() =>
            <CountAggregationOverlay count={count} link={link} items={items} Row={Row} />}
    >
        <Link href={link}>{count}</Link>
    </Overlay.Trigger>;

export { _CountAggregationOverlay };
export default CountAggregationCell;
