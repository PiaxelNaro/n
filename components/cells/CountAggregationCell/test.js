// @flow
import React from 'react';
import { shallow } from 'enzyme';
import CountAggregationCell, { _CountAggregationOverlay as CountAggregationOverlay } from '.';
import mockI18n from '../../../modules/i18n/mockI18n';

describe('<CountAggregationCell />', () => {
    it('renders overlay trigger', () => {
        const FakeRow: any = jest.fn();
        FakeRow.displayName = 'FakeRow';
        const component = shallow(
            <CountAggregationCell count={3} link="https://foo" items={[]} Row={FakeRow} />
        );
        expect(component).toMatchSnapshot();
        const overlay = shallow(component.find('OverlayTrigger').prop('getOverlay')());
        expect(overlay).toMatchSnapshot();
    });
});

describe('<CountAggregationOverlay />', () => {
    it('renders small list', () => {
        const FakeRow: any = jest.fn();
        FakeRow.displayName = 'FakeRow';
        const component = shallow(
            <CountAggregationOverlay
                i18n={mockI18n}
                count={3}
                link="https://foo"
                items={[
                    {
                        name: 'bar',
                    },
                    {
                        name: 'foo',
                    },
                ]}
                Row={FakeRow}
            />
        );
        expect(component).toMatchSnapshot();
    });

    it('renders see more link', () => {
        const MockRow: any = jest.fn();
        const component = shallow(
            <CountAggregationOverlay
                i18n={mockI18n}
                count={45}
                link="https://foo"
                items={[]}
                Row={MockRow}
            />
        );
        expect(component).toMatchSnapshot();
    });
});
