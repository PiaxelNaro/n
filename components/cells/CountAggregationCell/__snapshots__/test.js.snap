// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CountAggregationCell /> renders overlay trigger 1`] = `
<OverlayTrigger
  getOverlay={[Function]}
  trigger="hover"
>
  <CountAggregationCell__Link
    href="https://foo"
  >
    3
  </CountAggregationCell__Link>
</OverlayTrigger>
`;

exports[`<CountAggregationCell /> renders overlay trigger 2`] = `
<_CountAggregationOverlay
  Row={[Function]}
  count={3}
  items={Array []}
  link="https://foo"
/>
`;

exports[`<CountAggregationOverlay /> renders see more link 1`] = `
<Overlay
  showArrow={true}
>
  <CountAggregationCell__ListItem>
    <CountAggregationCell__Link
      href="https://foo"
    >
      + 45 more
    </CountAggregationCell__Link>
  </CountAggregationCell__ListItem>
</Overlay>
`;

exports[`<CountAggregationOverlay /> renders small list 1`] = `
<Overlay
  showArrow={true}
>
  <CountAggregationCell__ListItem>
    <FakeRow
      name="bar"
    />
  </CountAggregationCell__ListItem>
  <CountAggregationCell__ListItem>
    <FakeRow
      name="foo"
    />
  </CountAggregationCell__ListItem>
</Overlay>
`;
